### ISSUE TITLE
Hollow Metrics

### ISSUE THREAD

**Issue by rpalcolea** [Netflix/hollow#76]
Hi, 

Currently we’re running background jobs with quartz or simple future scheduling  on spring and ratpack to get metrics such as domain count object, heap used, refresh failed and so on and expose them via Prometheus. This has helped us to build grafana dashboards (screenshot attached)

<img width="1435" alt="screen shot 2017-08-06 at 8 36 52 pm" src="https://user-images.githubusercontent.com/1625920/29009286-9b8e592c-7ae7-11e7-9905-ad0d1504d4ec.png">

I'm opening this issue because I think it would be nice to have metrics out of the box from hollow and to see if the following  could fit on your design principles and contribute with it:

1. Introduce a `HollowMetrics` object as part of `HollowClientUpdater` to keep metrics for things such as domain object count by type (`readEngine.typeStates`), heap usage (`typeState.getApproximateHeapFootprintInBytes()`), current version, refresh succeeded/failed.  Every time a refresh fails or it's completed,  the metrics would be updated.

2. Introduce a `HollowMetricsCollector` interface: This could be part of the Consumer API ex. `HollowConsumer.withMetricsCollector(myMetricsCollector)`. Provide this as part of the `HollowClientUpdater` and invoke a method ex. `collect` when an update happens (refresh fails or succeeds - https://github.com/Netflix/hollow/blob/ecabde9e763ce4a7a6e128c86e3bf4c8e7a34f2c/hollow/src/main/java/com/netflix/hollow/api/client/HollowClientUpdater.java#L100 / https://github.com/Netflix/hollow/blob/ecabde9e763ce4a7a6e128c86e3bf4c8e7a34f2c/hollow/src/main/java/com/netflix/hollow/api/client/HollowClientUpdater.java#L104). This way, users could include their own mechanism to use those metrics such as writing to logs or exposing via JMX but that shouldn't be part of Hollow.

All of the above could be part of Hollow core, in addition to this I think it would be nice to have a `hollow-metrics` module (similar to `hollow-ui-explorer` and others). This module could have default implementations for `HollowMetricsCollector` using Micrometer (http://micrometer.io/)... think SLF4J, but for application metrics. Micrometer allows to expose metrics via Prometheus, datadog, atlas, influx, graphite and others.  This could be part of a second iteration.

your thoughts?

**Comment by dkoszewnik** [comment]
We think this is a good idea -- both the interfaces in core and the separate `hollow-metrics` project with default implementation.  We do have a couple of questions:

* You mentioned using background jobs to gather metrics.  Did you consider using a `RefreshListener` to trigger metrics gathering on consumer updates?  If so, curious about the rationale for not selecting this approach.
* Do you envision that the `HollowMetrics` is focused on consumer-side metrics?  If so, should it be called a `HollowConsumerMetrics`?  Further, would there be a corresponding `HollowProducerMetrics` which can also be used with the `HollowMetricsCollector`?

**Comment by rpalcolea** [comment]
Hi, great!

1. No, we created a common library across our services to use hollow consumers. That was ~3 months ago or probably more so it's a little outdated, we started to use the jobs on each API using the consumer back then. Now that we are moving to latest version of Hollow we thought it would be nice to have it inside, the idea of using the `RefreshListener` came to our minds a few days ago too.

2. That is a good point, I was thinking on consumer only but it could be also for producer

I'll work in a first version of this an open a PR
